src/Microsoft.AspNet.OData/Adapters/WebApiContext.cs

```diff
@@ -90,12 +91,41 @@ public string RouteName
         }
 
         /// <summary>
-        /// Gets or sets the parsed OData <see cref="SelectExpandClause"/> of the request.
+        /// Gets or sets the processed OData <see cref="SelectExpandClause"/> of the request. It will be different from QueryOptions.SelectExpand. SelectExpandClause as the $level is processed in this clause. 
         /// </summary>
-        public SelectExpandClause SelectExpandClause
+        public SelectExpandClause ProcessedSelectExpandClause
         {
-            get { return this.innerContext.SelectExpandClause; }
-            set { this.innerContext.SelectExpandClause = value; }
+            get
+            {
+                if (this.QueryOptions != null && this.QueryOptions.SelectExpand != null)
+                {
+                    SelectExpandClause selectExpandClause = this.QueryOptions.SelectExpand.ProcessedSelectExpandClause;
+                    if (selectExpandClause != null)
+                    {
+                        return selectExpandClause;
+                    }
+                }
+
+                return this.innerContext.SelectExpandClause;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the parsed OData <see cref="ODataQueryOptions"/> of the request.
+        /// </summary>
+        public ODataQueryOptions QueryOptions
+        {
+            get { return this.innerContext.QueryOptions; }
+            set
+            {
+                if (value != null && value.SelectExpand != null &&
+                    value.SelectExpand.ProcessedSelectExpandClause != null)
```

Comment:
if QueryOptions is null, or SelectExpand is null, or ProcessedSelectExpandClause is null, wouldn't the right behavior be to set ProcessedSelectExpandClause to null?  i.e., should we have an:
else
{
   this.innerContext.SelectExpandClause = null;
}